File: analysis/community-variability/matlab/scripts/ifcb_seasonal_comparison.m
%% IFCB seasonal comparison summaries in MATLAB
%
% Reads saved bootstrap, sensitivity, and spatial compositional variability
% outputs, then writes cross-season summary CSVs used by plotting workflows.

run("matlab/scripts/ifcb_common.m")

communityMetricOrder = ["CV_gamma", "CV_alpha", "CV_phi", "BD_gamma", "BD_alpha", "BD_phi", "BD_beta"];
plotVarsSensitivity = ["BD_gamma", "CV_gamma", "BD_phi", "CV_phi"];
topNTaxa = 10;

summaryAll = table();
bootAll = table();
for seasonFilter = seasons
    s = readtable(fullfile(resultsDir, "summary_" + seasonFilter + ".csv"));
    s.season = repmat(seasonFilter, height(s), 1);
    summaryAll = [summaryAll; s]; %#ok<AGROW>
    b = readtable(fullfile(resultsDir, "boot_" + seasonFilter + ".csv"));
    b.season = repmat(seasonFilter, height(b), 1);
    bootAll = [bootAll; b]; %#ok<AGROW>
end
bootAll = bootAll(string(bootAll.sample_type) == "Bootstrap", :);

summaryWide = unstack(summaryAll(:, ["season", "varname", "estimate"]), "estimate", "varname");
writetable(summaryWide, fullfile(resultsDir, "seasonal_metric_summary_wide.csv"));

looYearAll = table();
looTaxonAll = table();
for seasonFilter = seasons
    looYearAll = [looYearAll; readtable(fullfile(resultsDir, "leave_one_year_out_" + seasonFilter + ".csv"))]; %#ok<AGROW>
    looTaxonAll = [looTaxonAll; readtable(fullfile(resultsDir, "leave_one_taxon_out_" + seasonFilter + ".csv"))]; %#ok<AGROW>
end
looYearAll = add_baseline_delta_local(looYearAll, "year_removed", "season");
looTaxonAll = add_baseline_delta_local(looTaxonAll, "taxon_removed", "season");
writetable(looYearAll, fullfile(resultsDir, "leave_one_year_all_seasons_with_delta.csv"));
writetable(looTaxonAll, fullfile(resultsDir, "leave_one_taxon_all_seasons_with_delta.csv"));
writetable(classify_taxon_effects_local(looTaxonAll, plotVarsSensitivity, topNTaxa), ...
    fullfile(resultsDir, "leave_one_taxon_top_effects_all_seasons.csv"));

ellipsePairs = ["BD_alpha", "CV_alpha"; "BD_phi", "CV_phi"; "BD_gamma", "CV_gamma"; "BD_beta", "BD_gamma"];
ellipses = table();
for k = 1:size(ellipsePairs, 1)
    ellipses = [ellipses; make_ellipse_local(bootAll, seasons, ellipsePairs(k, 1), ellipsePairs(k, 2))]; %#ok<AGROW>
end
writetable(ellipses, fullfile(resultsDir, "bootstrap_metric_ellipses.csv"));

spavarAll = table();
for idx = 1:numel(seasons)
    seasonFilter = seasons(idx);
    s = readtable(fullfile(resultsDir, "spatial_variance_" + seasonFilter + ".csv"));
    s.season = repmat(seasonFilter, height(s), 1);
    s.year = str2double(string(s.timestep));
    s.time_index = s.year + (idx - 1) / numel(seasons);
    spavarAll = [spavarAll; s]; %#ok<AGROW>
end
writetable(spavarAll, fullfile(resultsDir, "spatial_variance_all_seasons.csv"));

function out = add_baseline_delta_local(dat, removedCol, groupCol)
out = dat;
out.baseline = nan(height(out), 1);
out.delta = nan(height(out), 1);
out.abs_delta = nan(height(out), 1);
for i = 1:height(out)
    sameMetric = string(out.varname) == string(out.varname(i));
    sameGroup = string(out.(groupCol)) == string(out.(groupCol)(i));
    baselineRow = sameMetric & sameGroup & string(out.(removedCol)) == "Baseline";
    if any(baselineRow)
        baselineValue = out.estimate(find(baselineRow, 1));
        out.baseline(i) = baselineValue;
        out.delta(i) = out.estimate(i) - baselineValue;
        out.abs_delta(i) = abs(out.delta(i));
    end
end
end

function out = classify_taxon_effects_local(dat, plotVars, topN)
rows = table();
for season = unique(string(dat.season))'
    for varname = plotVars
        d = dat(string(dat.season) == season & string(dat.varname) == varname & string(dat.taxon_removed) ~= "Baseline", :);
        if height(d) == 0; continue; end
        q = quantile(d.delta, [0.05, 0.95]);
        d.effect_class = repmat("Not significant", height(d), 1);
        d.effect_class(d.delta < q(1)) = "Negative";
        d.effect_class(d.delta > q(2)) = "Positive";
        d = sortrows(d, "abs_delta", "descend");
        rows = [rows; d(1:min(topN, height(d)), :)]; %#ok<AGROW>
    end
end
out = rows;
end

function out = make_ellipse_local(dat, seasons, xvar, yvar)
theta = linspace(0, 2*pi, 200)';
r2 = 5.991464547107979;
out = table();
for season = seasons
    d = dat(string(dat.season) == season, :);
    xy = [d.(xvar), d.(yvar)];
    xy = xy(all(isfinite(xy), 2), :);
    if size(xy, 1) < 3; continue; end
    S = cov(xy);
    if any(~isfinite(S), "all") || det(S) <= 0; continue; end
    mu = mean(xy, 1);
    [V, D] = eig(S);
    A = V * diag(sqrt(max(diag(D), 0) * r2));
    pts = mu + [cos(theta), sin(theta)] * A';
    tmp = table(repmat(season, size(pts, 1), 1), pts(:, 1), pts(:, 2), repmat(xvar, size(pts, 1), 1), repmat(yvar, size(pts, 1), 1), ...
        "VariableNames", ["season", "x", "y", "xvar", "yvar"]);
    out = [out; tmp]; %#ok<AGROW>
end
end

File: analysis/community-variability/scripts/ifcb_seasonal_comparison.py
"""IFCB seasonal comparison summaries in Python."""

from __future__ import annotations

import argparse
from pathlib import Path

import numpy as np
import pandas as pd

from ifcb_common import SEASONS, default_results_dir
from ifcb.neslter.community_variability import COMMUNITY_METRIC_ORDER, add_baseline_delta


PLOT_VARS_SENSITIVITY = ["BD_gamma", "CV_gamma", "BD_phi", "CV_phi"]


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Prepare seasonal comparison CSVs from Python analysis outputs.")
    parser.add_argument("--results-dir", default=default_results_dir(), type=str)
    parser.add_argument("--seasons", nargs="+", default=SEASONS)
    parser.add_argument("--top-n-taxa", default=10, type=int)
    return parser.parse_args()


def read_seasonal_csv(results_dir: Path, seasons: list[str], prefix: str) -> pd.DataFrame:
    rows = []
    for season in seasons:
        path = results_dir / f"{prefix}_{season}.csv"
        data = pd.read_csv(path)
        if "season" not in data.columns:
            data.insert(0, "season", season)
        rows.append(data)
    return pd.concat(rows, ignore_index=True)


def make_ellipse(data: pd.DataFrame, seasons: list[str], xvar: str, yvar: str) -> pd.DataFrame:
    """Calculate 95% covariance ellipse points: (x - mu)' S^-1 (x - mu) = chi2_0.95."""
    theta = np.linspace(0.0, 2.0 * np.pi, 200)
    r2 = 5.991464547107979  # qchisq(0.95, df = 2)
    rows = []
    for season in seasons:
        d = data.loc[data["season"] == season, [xvar, yvar]].dropna()
        if len(d) < 3:
            continue
        cov = np.cov(d.to_numpy(dtype=float), rowvar=False, ddof=1)
        if not np.all(np.isfinite(cov)) or np.linalg.det(cov) <= 0:
            continue
        mu = d.mean(axis=0).to_numpy(dtype=float)
        eigval, eigvec = np.linalg.eigh(cov)
        transform = eigvec @ np.diag(np.sqrt(np.maximum(eigval, 0.0) * r2))
        points = mu + np.column_stack([np.cos(theta), np.sin(theta)]) @ transform.T
        rows.append(pd.DataFrame({"season": season, "x": points[:, 0], "y": points[:, 1], "xvar": xvar, "yvar": yvar}))
    return pd.concat(rows, ignore_index=True) if rows else pd.DataFrame(columns=["season", "x", "y", "xvar", "yvar"])


def classify_taxon_effects(loo_taxon_all: pd.DataFrame, top_n_taxa: int) -> pd.DataFrame:
    """Classify taxon removals using seasonal 5th and 95th percentile delta thresholds."""
    thresholds = (
        loo_taxon_all.loc[loo_taxon_all["taxon_removed"] != "Baseline"]
        .groupby(["season", "varname"])["delta"]
        .quantile([0.05, 0.95])
        .unstack()
        .rename(columns={0.05: "lwr", 0.95: "upr"})
        .reset_index()
    )
    data = loo_taxon_all.loc[
        (loo_taxon_all["taxon_removed"] != "Baseline")
        & loo_taxon_all["varname"].isin(PLOT_VARS_SENSITIVITY)
    ].merge(thresholds, on=["season", "varname"], how="left")

    # Positive means removing the taxon increases the metric; negative means it decreases it.
    significant = (data["delta"] < data["lwr"]) | (data["delta"] > data["upr"])
    data["effect_class"] = "Not significant"
    data.loc[significant & (data["delta"] > 0), "effect_class"] = "Positive"
    data.loc[significant & (data["delta"] < 0), "effect_class"] = "Negative"

    return (
        data.sort_values(["season", "varname", "abs_delta"], ascending=[True, True, False])
        .groupby(["season", "varname"], as_index=False)
        .head(top_n_taxa)
        .reset_index(drop=True)
    )


def main() -> int:
    args = parse_args()
    results_dir = Path(args.results_dir)
    seasons = list(args.seasons)

    # Read bootstrap summaries and replicate-level metric tables.
    summary_all = read_seasonal_csv(results_dir, seasons, "summary")
    boot_all = read_seasonal_csv(results_dir, seasons, "boot")
    boot_all = boot_all.loc[boot_all["sample_type"] == "Bootstrap"].copy()

    summary_all["varname"] = pd.Categorical(summary_all["varname"], categories=COMMUNITY_METRIC_ORDER, ordered=True)
    summary_wide = summary_all.pivot(index="season", columns="varname", values=["estimate", "lwr", "upr"])
    summary_wide.columns = [f"{metric}_{value}" for value, metric in summary_wide.columns]
    summary_wide = summary_wide.reset_index()
    summary_wide.to_csv(results_dir / "seasonal_metric_summary_wide.csv", index=False)

    # Leave-one-year and leave-one-taxon deltas: delta = metric_without_component - baseline.
    loo_year_all = read_seasonal_csv(results_dir, seasons, "leave_one_year_out")
    loo_year_all = add_baseline_delta(loo_year_all, removed_col="year_removed", group_cols="season")
    loo_year_all.to_csv(results_dir / "leave_one_year_all_seasons_with_delta.csv", index=False)

    loo_taxon_all = read_seasonal_csv(results_dir, seasons, "leave_one_taxon_out")
    loo_taxon_all = add_baseline_delta(loo_taxon_all, removed_col="taxon_removed", group_cols="season")
    loo_taxon_all.to_csv(results_dir / "leave_one_taxon_all_seasons_with_delta.csv", index=False)
    classify_taxon_effects(loo_taxon_all, args.top_n_taxa).to_csv(
        results_dir / "leave_one_taxon_top_effects_all_seasons.csv",
        index=False,
    )

    # Bootstrap metric-space ellipses used by the R seasonal comparison figures.
    ellipse_pairs = [
        ("BD_alpha", "CV_alpha"),
        ("BD_phi", "CV_phi"),
        ("BD_gamma", "CV_gamma"),
        ("BD_beta", "BD_gamma"),
    ]
    ellipses = [make_ellipse(boot_all, seasons, xvar, yvar) for xvar, yvar in ellipse_pairs]
    pd.concat(ellipses, ignore_index=True).to_csv(results_dir / "bootstrap_metric_ellipses.csv", index=False)

    # Time-resolved spatial compositional variability: BD_t^h = sum_j Var_i(z_itj).
    spavar_all = read_seasonal_csv(results_dir, seasons, "spatial_variance")
    spavar_all["year"] = pd.to_numeric(spavar_all["timestep"], errors="coerce")
    spavar_all["season_index"] = spavar_all["season"].map({season: i for i, season in enumerate(seasons)})
    spavar_all["time_index"] = spavar_all["year"] + spavar_all["season_index"] / len(seasons)
    spavar_all.to_csv(results_dir / "spatial_variance_all_seasons.csv", index=False)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
